### ISSUE TITLE
Question: Is it possible to get args params for an static void main(String[] args)?

### ISSUE THREAD

**Issue by EdwinFLopez** [groovy/GMavenPlus#304]
Hi.

First of all, thanks for this plugin. I have a bit of a problem, it is not a blocker, though. It is a question: is it possible to obtain the passed `args` when running a script with a `static void main` method? If so, how to?

When I execute the groovy script as a standalone, it works, no issue, I get the args. The problem is when I use the plugin execute command: `mvn gplus:execute -Dargs=blah -DanotherParam=one`

I get the following error:

```bash
[ERROR] Failed to execute goal org.codehaus.gmavenplus:gmavenplus-plugin:3.0.0:execute (default-cli) on project migration: Error occurred while calling a method on a Groovy class from classpath.: InvocationTargetException: No such property: args for class: migration -> [Help 1]
[ERROR] 
```
Is it possible that I'm doing something wrongly? Thank you very much.

I have the following pom.xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>migration</groupId>
    <artifactId>migration</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>org.apache.groovy</groupId>
            <artifactId>groovy-all</artifactId>
            <version>${groovy-all.version}</version>
            <type>pom</type>
        </dependency>
    </dependencies>
    <build>
        <plugins>
            <plugin>
                <groupId>org.codehaus.gmavenplus</groupId>
                <artifactId>gmavenplus-plugin</artifactId>
                <version>${gmavenplus-plugin.version}</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>execute</goal>
                        </goals>
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>org.apache.groovy</groupId>
                        <artifactId>groovy</artifactId>
                        <version>${groovy.version}</version>
                        <scope>runtime</scope>
                    </dependency>
                </dependencies>
                <configuration>
                    <warningLevel>3</warningLevel>
                    <verbose>true</verbose>
                    <targetBytecode>${java.version}</targetBytecode>
                    <scripts>
                        <script>${project.basedir}/src/main/groovy/migration.groovy</script>
                    </scripts>
                    <properties>
                        <property>
                            <name>TARGET_ENVIRONMENT</name>
                            <value>${env.TARGET_ENVIRONMENT}</value>
                        </property>
                        <property>
                            <name>CONFIG_LOCATION</name>
<value>${project.basedir}/src/main/resources/migration.config</value>
                        </property>
                    </properties>
                </configuration>
            </plugin>
        </plugins>
    </build>

    <properties>
        <gmavenplus-plugin.version>3.0.0</gmavenplus-plugin.version>
        <groovy.version>4.0.14</groovy.version>
        <groovy-all.version>${groovy.version}</groovy-all.version>
        <java.version>17</java.version>
        <maven.compiler.source>${java.version}</maven.compiler.source>
        <maven.compiler.target>${java.version}</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>
</project>
```

and the groovy script:

```groovy
#!/usr/bin/env groovy

static void main(String[] args) {
    def sysprops = new Properties()
    def addProperty = { k, v -> sysprops.setProperty(k, v) }
    System.properties.each(addProperty)
    System.getenv().each(addProperty)

    println("keys: ${sysprops.keySet().sort()}")
    println("target-environment: ${sysprops['TARGET_ENVIRONMENT']}")
    println("config-location: ${sysprops['CONFIG_LOCATION']}")
    println("target-environment-1: ${TARGET_ENVIRONMENT}")
    println("config-location-1: ${CONFIG_LOCATION}")
    println("args: ${args}")
}
```

**Comment by keeganwitt** [comment]
There are a couple things to understand on this. First, the name `args` is arbitrary as far as Java/Groovy is concerned. You could call that parameter anything you like in a main function. Second, the properties you are setting via command line are setting system properties as documented in the Maven book [here](https://books.sonatype.com/mvnref-book/reference/running-sect-options.html#running-sect-define-prop).

I can look into the possibility of setting main arguments. But it would be done via a newly introduce Maven parameter (which can be set via command line or in POM). Is that something that'd be useful?

In the meantime, what you'd have to do is check for both possibilities in your script, like this

```groovy
static void main(def args) {
    def mainParamaterArguments = args
    if (args.size() == 0) {
        mainParamaterArguments = mainArguments.split(",")  // or whatever name you want for the new Maven parameter
        // GMavenPlus handles binding `mainArguments` from the Maven system property, but you can also do that explicitly like below
       // mainParamaterArguments = System.properties["mainArguments"].split(",")
    }
}
```

**Comment by EdwinFLopez** [comment]
Super fast answer! Thank you so much.

> There are a couple things to understand on this. First, the name args is arbitrary as far as Java/Groovy is concerned. You could call that parameter anything you like in a main function. Second, the properties you are setting via command line are setting system properties as documented in the Maven book [here](https://books.sonatype.com/mvnref-book/reference/running-sect-options.html#running-sect-define-prop).
Yes, it was a poor and confusing example, my apologies, I'm totally aware of this. Thanks for pointing it out.

> Is that something that'd be useful?
Not so sure, since I believe that maven will swallow the params, therefore setting a new maven goal parameter seems the only way to go.

The expectation, I believe, is that, whether the script is called via command-shell or maven goal, the standard `void main(String... args)` method should receive the scripts params.

To sum up, this answers my question: there's no current way to pass the execution params.

I believe that this is such a small thing with not that many use cases not worth the effort of a new release, maybe a documentation update.

I'll close the case since I'm satisfied 😃 . Thank you very much for such a prompt answer. Have the greatest of days ahead.
